### ISSUE TITLE
"if not null" operators

### ISSUE THREAD

**Issue by har79** [dart-lang/language#131]
Dart supports "if null" access and assignment operators — `??`, `??=`.

Equivalent "if not null" operators would also be useful, particularly when non-nullable types land (https://github.com/dart-lang/language/issues/110). They would allow concise use of nullable variables in a similar manner to quiver's `Optional`. E.g., instead of:

```
Optional<Foo> foo;
foo = foo.transform(changeFoo);
final bar = foo.transform(fooToBar);
```

nullable variables could be used:

```
Foo? foo;
foo ?!= changeFoo(foo);
final bar = foo ?! fooToBar(foo);
```

**Comment by lrhn** [comment]
It is a common pattern to "do something if something is not null". The `?.` operation allows calling *methods* on an object if it is not null, but there is no similar feature to guard a general operation by something being not-null (the opposite of `??`).

The `?!` operator is probably not viable (the sequence can occur naturally, e.g., `x ?! b : c`).

**Comment by har79** [comment]
Good point, `!?` and `!?=` should be a more viable alternative and more memorable since it matches the structure of the long-form equivalent: `x ! /*= null*/ ? b /*: null*/`

**Comment by DonizeteVida** [comment]
Something like that: 

border: borderColor ?? Border.all(color: borderColor, width: 1),

border is optional, so I have to do: 

border: borderColor != null
                ? Border.all(color: borderColor, width: 1)
                : null,


border: borderColor !? Border.all(color: borderColor, width: 1),

**Comment by tatumizer** [comment]
If anonymous methods (#260) are implemented, you would be able to write it as
```
border: borderColor?.{return Border.all(color: this, width: 1);}
```
or even something like this:
```
border: borderColor?.{-> Border.all(color: this, width: 1)}
```
This doesn't require adding any new operators b/c `expr?.method` might be enough to convey the meaning of "if expr is not null, do this".

NOTE: It seems we have two concurrent proposals competing for symbol `->` : #260 and #762. Maybe we can find a way to reconcile them?

**Comment by llucax** [comment]
Probably not very compatible with Dart explicit bool rules, but another option would be to allow an exception with null and use the `&&` short-circuit operator like in C-like languages: `border: borderColor && Border.all(color: boderColor, width: 1)`.

**Comment by mabdelaal86** [comment]
I just created an extension function as follows:
```dart
extension NotNullExtension<T> on T {
  R convert<R>(R Function(T e) fun) => fun(this);
}
```

So, I can call it as `y = x?.convert((v) => v + 1) ?? 0` for example.

The following is a real example:
```dart
class IconButton extends Button {
    Button({String? text, String? icon})
    : super(text: icon?.codePoint.convert(String.fromCharCode) ?? text);
}
```

It would be nice to have a similar built-in function or operator.

**Comment by FMorschel** [comment]
What about a new string interpolation one?

```dart
void foo(int? i) => '$?i';
```

This would print either a number or an empty string instead of a number or `'null'`.

**Comment by lrhn** [comment]
String interpolation elements, [yes please](https://github.com/dart-lang/language/issues/1479).

A null-aware identifier element without braces should be possible.
